Error downloading driver libraries

### Description

I updated to 23.3.1.202312241705 today and noticed in the release notes that there was a newer Oracle library version being used.  When DBeaver started, it attempted to download the libraries but I receive an error message (seen below).

> Error downloading driver libraries
> 
> Reason:
> Error resolving dependencies
>   Index 0 out of bounds for length 0

<details>
<summary>Log from .log:</summary>

```
!ENTRY org.jkiss.dbeaver.model 4 0 2023-12-26 13:23:21.481
!MESSAGE Index 0 out of bounds for length 0
!STACK 0
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Unknown Source)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Unknown Source)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Unknown Source)
	at java.base/java.util.Objects.checkIndex(Unknown Source)
	at java.base/java.util.ArrayList.get(Unknown Source)
	at org.jkiss.dbeaver.registry.maven.MavenArtifact.getFileURL(MavenArtifact.java:309)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.getRemotePOMLocation(MavenArtifactVersion.java:252)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.cachePOM(MavenArtifactVersion.java:259)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.loadPOM(MavenArtifactVersion.java:277)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.<init>(MavenArtifactVersion.java:101)
	at org.jkiss.dbeaver.registry.maven.MavenArtifact.makeLocalVersion(MavenArtifact.java:345)
	at org.jkiss.dbeaver.registry.maven.MavenArtifact.resolveVersion(MavenArtifact.java:424)
	at org.jkiss.dbeaver.registry.maven.MavenRepository.findArtifact(MavenRepository.java:204)
	at org.jkiss.dbeaver.registry.maven.MavenRegistry.findInRepositories(MavenRegistry.java:226)
	at org.jkiss.dbeaver.registry.maven.MavenRegistry.findArtifact(MavenRegistry.java:185)
	at org.jkiss.dbeaver.registry.driver.DriverLibraryMavenArtifact.getArtifactVersion(DriverLibraryMavenArtifact.java:202)
	at org.jkiss.dbeaver.registry.driver.DriverLibraryMavenArtifact.resolveLocalVersion(DriverLibraryMavenArtifact.java:339)
	at org.jkiss.dbeaver.registry.driver.DriverLibraryMavenArtifact.getDependencies(DriverLibraryMavenArtifact.java:247)
	at org.jkiss.dbeaver.registry.driver.DriverDependencies.resolveDependencies(DriverDependencies.java:107)
	at org.jkiss.dbeaver.registry.driver.DriverDependencies.resolveDependencies(DriverDependencies.java:57)
	at org.jkiss.dbeaver.ui.dialogs.driver.DriverDependenciesTree.lambda$1(DriverDependenciesTree.java:130)
	at org.jkiss.dbeaver.runtime.RunnableContextDelegate.lambda$0(RunnableContextDelegate.java:39)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)

!ENTRY org.jkiss.dbeaver.model 2 0 2023-12-26 13:24:51.989
!MESSAGE Can't load driver properties
!SUBENTRY 1 org.jkiss.dbeaver.model 2 0 2023-12-26 13:24:51.989
!MESSAGE Error creating driver 'Oracle' instance.
Most likely required jar files are missing.
You should configure jars in driver settings.

Reason: can't load driver class 'oracle.jdbc.OracleDriver'
!SUBENTRY 2 org.jkiss.dbeaver.model 2 0 2023-12-26 13:24:51.989
!MESSAGE oracle.jdbc.OracleDriver
!STACK 0
java.lang.ClassNotFoundException: oracle.jdbc.OracleDriver
	at java.base/java.net.URLClassLoader.findClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Unknown Source)
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.loadDriver(DriverDescriptor.java:1243)
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.loadDriver(DriverDescriptor.java:1176)
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.getDriverInstance(DriverDescriptor.java:716)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSourceProvider.getConnectionProperties(JDBCDataSourceProvider.java:63)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionPropertiesControl.loadDriverProperties(ConnectionPropertiesControl.java:164)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionPropertiesControl.makeProperties(ConnectionPropertiesControl.java:85)
	at org.jkiss.dbeaver.ui.dialogs.connection.DriverPropertiesDialogPage.lambda$0(DriverPropertiesDialogPage.java:89)
	at org.jkiss.dbeaver.runtime.RunnableContextDelegate.lambda$0(RunnableContextDelegate.java:39)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
```
</details>

<details>
<summary>Log from dbeaver-debug.log:</summary>

```
2023-12-26 13:23:19.427 - Service 'org.jkiss.dbeaver.runtime.ui.UIServiceSecurity' not found
2023-12-26 13:23:21.229 - Open [https://repo1.maven.org]
2023-12-26 13:23:21.472 - Index 0 out of bounds for length 0
java.lang.IndexOutOfBoundsException: Index 0 out of bounds for length 0
	at java.base/jdk.internal.util.Preconditions.outOfBounds(Unknown Source)
	at java.base/jdk.internal.util.Preconditions.outOfBoundsCheckIndex(Unknown Source)
	at java.base/jdk.internal.util.Preconditions.checkIndex(Unknown Source)
	at java.base/java.util.Objects.checkIndex(Unknown Source)
	at java.base/java.util.ArrayList.get(Unknown Source)
	at org.jkiss.dbeaver.registry.maven.MavenArtifact.getFileURL(MavenArtifact.java:309)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.getRemotePOMLocation(MavenArtifactVersion.java:252)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.cachePOM(MavenArtifactVersion.java:259)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.loadPOM(MavenArtifactVersion.java:277)
	at org.jkiss.dbeaver.registry.maven.MavenArtifactVersion.<init>(MavenArtifactVersion.java:101)
	at org.jkiss.dbeaver.registry.maven.MavenArtifact.makeLocalVersion(MavenArtifact.java:345)
	at org.jkiss.dbeaver.registry.maven.MavenArtifact.resolveVersion(MavenArtifact.java:424)
	at org.jkiss.dbeaver.registry.maven.MavenRepository.findArtifact(MavenRepository.java:204)
	at org.jkiss.dbeaver.registry.maven.MavenRegistry.findInRepositories(MavenRegistry.java:226)
	at org.jkiss.dbeaver.registry.maven.MavenRegistry.findArtifact(MavenRegistry.java:185)
	at org.jkiss.dbeaver.registry.driver.DriverLibraryMavenArtifact.getArtifactVersion(DriverLibraryMavenArtifact.java:202)
	at org.jkiss.dbeaver.registry.driver.DriverLibraryMavenArtifact.resolveLocalVersion(DriverLibraryMavenArtifact.java:339)
	at org.jkiss.dbeaver.registry.driver.DriverLibraryMavenArtifact.getDependencies(DriverLibraryMavenArtifact.java:247)
	at org.jkiss.dbeaver.registry.driver.DriverDependencies.resolveDependencies(DriverDependencies.java:107)
	at org.jkiss.dbeaver.registry.driver.DriverDependencies.resolveDependencies(DriverDependencies.java:57)
	at org.jkiss.dbeaver.ui.dialogs.driver.DriverDependenciesTree.lambda$1(DriverDependenciesTree.java:130)
	at org.jkiss.dbeaver.runtime.RunnableContextDelegate.lambda$0(RunnableContextDelegate.java:39)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
2023-12-26 13:24:51.986 - Can't load driver properties
org.jkiss.dbeaver.DBException: Error creating driver 'Oracle' instance.
Most likely required jar files are missing.
You should configure jars in driver settings.

Reason: can't load driver class 'oracle.jdbc.OracleDriver'
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.loadDriver(DriverDescriptor.java:1245)
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.loadDriver(DriverDescriptor.java:1176)
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.getDriverInstance(DriverDescriptor.java:716)
	at org.jkiss.dbeaver.model.impl.jdbc.JDBCDataSourceProvider.getConnectionProperties(JDBCDataSourceProvider.java:63)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionPropertiesControl.loadDriverProperties(ConnectionPropertiesControl.java:164)
	at org.jkiss.dbeaver.ui.dialogs.connection.ConnectionPropertiesControl.makeProperties(ConnectionPropertiesControl.java:85)
	at org.jkiss.dbeaver.ui.dialogs.connection.DriverPropertiesDialogPage.lambda$0(DriverPropertiesDialogPage.java:89)
	at org.jkiss.dbeaver.runtime.RunnableContextDelegate.lambda$0(RunnableContextDelegate.java:39)
	at org.eclipse.jface.operation.ModalContext$ModalContextThread.run(ModalContext.java:122)
Caused by: java.lang.ClassNotFoundException: oracle.jdbc.OracleDriver
	at java.base/java.net.URLClassLoader.findClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.ClassLoader.loadClass(Unknown Source)
	at java.base/java.lang.Class.forName0(Native Method)
	at java.base/java.lang.Class.forName(Unknown Source)
	at org.jkiss.dbeaver.registry.driver.DriverDescriptor.loadDriver(DriverDescriptor.java:1243)
	... 8 more
```
</details>

### DBeaver Version

23.3.1.202312241705

### Operating System

Windows 10.0.19045.3803

### Database and driver

oracle.jdbc.OracleDriver
23.2.0.0

### Steps to reproduce

Attempt to connect to an Oracle database without the necessary drivers installed yet.

### Additional context

I thought maybe it was a proxy issue but entering credentials for my organization's proxy didn't improve the situation.

--- a/plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/maven/MavenArtifact.java
+++ b/plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/maven/MavenArtifact.java
@@ -341,7 +341,13 @@ public class MavenArtifact implements IMavenIdentifier
         return null;
     }
 
-    private MavenArtifactVersion makeLocalVersion(DBRProgressMonitor monitor, String versionStr, boolean setActive, boolean resolveOptionalDependencies, boolean snapshotVersion) throws IllegalArgumentException, IOException {
+    private MavenArtifactVersion makeLocalVersion(
+        DBRProgressMonitor monitor,
+        String versionStr,
+        boolean setActive,
+        boolean resolveOptionalDependencies,
+        boolean snapshotVersion
+    ) throws IllegalArgumentException, IOException {
         MavenArtifactVersion version = getVersion(versionStr);
         if (version == null) {
             version = new MavenArtifactVersion(monitor, this, versionStr, resolveOptionalDependencies, snapshotVersion);

--- a/plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/maven/MavenArtifact.java
+++ b/plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/maven/MavenArtifact.java
@@ -70,7 +70,7 @@ public class MavenArtifact implements IMavenIdentifier
     private String latestVersion;
     private String releaseVersion;
     private Date lastUpdate;
-    private List<String> snapshotVersions = new ArrayList<>();
+    private final List<String> snapshotVersions = new ArrayList<>();
     private final List<MavenArtifactVersion> localVersions = new ArrayList<>();
 
     private transient boolean metadataLoaded = false;
@@ -106,7 +106,7 @@ public class MavenArtifact implements IMavenIdentifier
             metadataPath += version + "/";
         }
         metadataPath += MAVEN_METADATA_XML;
-        monitor.subTask("Load metadata " + this + "");
+        monitor.subTask("Load metadata " + this);
 
         try (InputStream mdStream = WebUtils.openConnection(metadataPath, getRepository().getAuthInfo(), null).getInputStream()) {
             parseMetadata(mdStream);
@@ -304,10 +304,12 @@ public class MavenArtifact implements IMavenIdentifier
         return repository.getUrl() + dir + "/";
     }
 
-    String getFileURL(String version, String fileType) {
-        if (getRepository().isSnapshot()) {
-            return getBaseArtifactURL() + versions.get(0) + "/" + getVersionFileName(VersionUtils.findLatestVersion(snapshotVersions),
-                fileType);
+    String getFileURL(String version, String fileType, boolean snapshotVersion) {
+        if (snapshotVersion) {
+            return getBaseArtifactURL() + versions.get(0) + "/" + getVersionFileName(
+                VersionUtils.findLatestVersion(snapshotVersions),
+                fileType
+            );
         }
 
         return getBaseArtifactURL() + version + "/" + getVersionFileName(version, fileType);
@@ -339,10 +341,10 @@ public class MavenArtifact implements IMavenIdentifier
         return null;
     }
 
-    private MavenArtifactVersion makeLocalVersion(DBRProgressMonitor monitor, String versionStr, boolean setActive, boolean resolveOptionalDependencies) throws IllegalArgumentException, IOException {
+    private MavenArtifactVersion makeLocalVersion(DBRProgressMonitor monitor, String versionStr, boolean setActive, boolean resolveOptionalDependencies, boolean snapshotVersion) throws IllegalArgumentException, IOException {
         MavenArtifactVersion version = getVersion(versionStr);
         if (version == null) {
-            version = new MavenArtifactVersion(monitor, this, versionStr, resolveOptionalDependencies);
+            version = new MavenArtifactVersion(monitor, this, versionStr, resolveOptionalDependencies, snapshotVersion);
             localVersions.add(version);
         }
         return version;
@@ -384,6 +386,9 @@ public class MavenArtifact implements IMavenIdentifier
                     break;
                 default:
                     if (snapshotVersion) {
+                        if (snapshotVersions.isEmpty()) {
+                            throw new IOException("Artifact '" + this + "' has empty snapshot version list");
+                        }
                         versionInfo = VersionUtils.findLatestVersion(snapshotVersions);
                         break;
                     }
@@ -421,7 +426,7 @@ public class MavenArtifact implements IMavenIdentifier
         MavenArtifactVersion localVersion = getVersion(versionInfo);
         if (localVersion == null) {
             try {
-                localVersion = makeLocalVersion(monitor, versionInfo, lookupVersion, resolveOptionalDependencies);
+                localVersion = makeLocalVersion(monitor, versionInfo, lookupVersion, resolveOptionalDependencies, snapshotVersion);
             } catch (IOException e) {
                 // Some IO error - not fatal
                 log.debug("Error loading version info: " + e.getMessage());

--- a/plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/maven/MavenArtifactVersion.java
+++ b/plugins/org.jkiss.dbeaver.registry/src/org/jkiss/dbeaver/registry/maven/MavenArtifactVersion.java
@@ -50,6 +50,7 @@ public class MavenArtifactVersion implements IMavenIdentifier {
     private static final String DEFAULT_PROFILE_ID = "#root";
 
     private final MavenArtifact artifact;
+    private final boolean snapshotVersion;
     private String name;
     private String version;
     private String packaging;
@@ -93,11 +94,12 @@ public class MavenArtifactVersion implements IMavenIdentifier {
         @NotNull DBRProgressMonitor monitor,
         @NotNull MavenArtifact artifact,
         @NotNull String version,
-        boolean resolveOptionalDependencies
+        boolean resolveOptionalDependencies,
+        boolean snapshotVersion
     ) throws IOException {
         this.artifact = artifact;
         this.version = CommonUtils.trim(version);
-
+        this.snapshotVersion = snapshotVersion;
         loadPOM(monitor, resolveOptionalDependencies);
         this.version = evaluateString(this.version);
 
@@ -206,7 +208,7 @@ public class MavenArtifactVersion implements IMavenIdentifier {
     }
 
     public String getExternalURL() {
-        return artifact.getFileURL(version, getPackagingFileExtension());
+        return artifact.getFileURL(version, getPackagingFileExtension(), snapshotVersion);
     }
 
     @NotNull
@@ -224,7 +226,7 @@ public class MavenArtifactVersion implements IMavenIdentifier {
     }
 
     public String getExternalURL(String fileType) {
-        return artifact.getFileURL(version, fileType);
+        return artifact.getFileURL(version, fileType, snapshotVersion);
     }
 
     public String getPath() {
@@ -249,7 +251,7 @@ public class MavenArtifactVersion implements IMavenIdentifier {
     }
 
     private String getRemotePOMLocation() {
-        return artifact.getFileURL(version, MavenArtifact.FILE_POM);
+        return artifact.getFileURL(version, MavenArtifact.FILE_POM, snapshotVersion);
     }
 
     private void cachePOM(File localPOM) throws IOException {

